s390x: Move the value out of the casloop_val_reg with mov_preg

elliottt opened the pull request:
The `casloop_emit` function in the s390x backend was using the fixed non-allocatable register `%r0` directly with move instructions, which produced a panic in the regalloc2 checker (#5425). This PR changes the `casloop_result` function to use `mov_preg` instead of `copy_reg` to fetch the result, as it's not viewed by regalloc2 as a move.

Fixes #5425 
<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

cfallin commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -568,13 +568,21 @@ impl<I: VCodeInst> VCodeBuilder<I> {
-                self.vcode.is_move.insert(InsnIndex::new(i), (src, dst));
+                // As we only use real registers with reg_fixed_def/reg_fixed_use constraints, the
+                // only situation where we would find a real register here is if it came from a
+                // reg_fixed_nonallocatable constraint. Don't add the alias at this point, as that
```
s/alias/move/ -- "aliases" are what are used to make one vreg resolve to another (so that result vregs can be rewritten to the temp vregs returned by the toplevel lowering entry point), and imply a "timeless" equivalence, while "moves" are specific copy operations that occur in program flow. (Likewise the title of this PR/commit should change)

cfallin reviewed (approved):
LGTM, with nit below. Thanks for tracking this down!

I also think we should have an assert in RA2 to catch this case and explicitly disallow it; but that can happen off the critical path here.

elliottt commented:
Looks like this isn't quite solved yet: the s390x tests that are failing are due to not recording an entry in the `is_move` map of the vcode table for the move that puts the result of the `atomic_rmw` instruction into `%r2` before returning.

Attempting a local fix by recording `Operand::fixed_nonallocatable` constraints in `is_move` when appropriate instead of avoiding adding an entry at all hits a bug in regalloc2 where we're assuming that a fixed constraint returned by `is_move` must always be a pinned vreg, and not a fixed-nonallocatable register.

elliottt commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -910,7 +918,7 @@ impl<I: VCodeInst> VCode<I> {
                             }
                         }
 
-                        if self.insts[iix.index()].is_move().is_some() {
+                        if self.is_move.contains_key(&iix) {
```
Reusing the cached decision about what instruction is a move meant that we keep the move instructions that involve fixed non-allocatable registers in the final assembly.

elliottt commented on cranelift/filetests/filetests/isa/s390x/atomic_rmw.clif:
```diff
@@ -573,7 +573,7 @@ block0(v0: i64, v1: i64, v2: i32):
 ;   l %r0, 0(%r3)
 ;   0: clr %r4, %r0 ; jgnh 1f ; cs %r0, %r4, 0(%r3) ; jglh 0b ; 1:
-;   lgr %r2, %r0
+;   lr %r2, %r0
```
@uweigand would you mind checking these move changes? They all seemed reasonable to me, as the result of each function is `i32`.

elliottt commented:
The last problem here was that moves weren't making it into the final output. The problem was that in the `VCode::emit` function we were skipping all instructions that satisfied `MachInst::is_move`, which includes the moves that we didn't report to regalloc2. The solution was to reuse the `VCode::is_move` map rather than calling `MachInst::is_move`, as presence in that map corresponds directly to whether or not we told regalloc2 about the move.

cfallin reviewed (changes requested):
Actually, reading your latest update: I think that what is actually needed here is for `MachInst::is_move()` to return a correct result with respect to real/virtual registers. Otherwise we have an inconsistency that may cause issues elsewhere (and are not addressed by papering over one callsite to `MachInst::is_move` with alternative logic).

elliottt commented:
@uweigand I've reworked this a bit more to avoid the changes to VCode. The problem was that there was a move using a fixed non-allocatable register produced by the s390x backend in the `casloop_emit` function. I've reworked this to use `MovPReg`, which is designed to handle fixed non-allocatable sources, but this does introduce some additional indirection for the result.

elliottt commented:
> Actually, reading your latest update: I think that what is actually needed here is for `MachInst::is_move()` to return a correct result with respect to real/virtual registers. Otherwise we have an inconsistency that may cause issues elsewhere (and are not addressed by papering over one callsite to `MachInst::is_move` with alternative logic).

The changes to `VCode` and `MachInst` ended up being a red-herring: the real problem as we discussed in a DM was that there were move instructions emitted that were using fixed non-allocatable registers at all. Switching to `MInst.MovPreg` instead in the s390x backend was the best fix.

cfallin reviewed (approved):
LGTM, thanks!

Just one nit before merging if you don't mind: could we add an assert in `collect_operands` (or somewhere else appropriate) where we invoke `is_move()` to check that the src and dest of any program move is a virtual register?

uweigand commented on cranelift/codegen/src/isa/s390x/inst.isle:
```diff
@@ -3424,17 +3427,17 @@
             (_ Unit (emit_load (ty_ext32 ty) (casloop_val_reg) aligned_mem))
             (_ Unit (emit_loop ib (intcc_as_cond (IntCC.NotEqual)))))
-        result))
+        (preg_gpr_0)))
```
This a bit iffy, in that it's not guaranteed (at least, that wasn't the intention) that `result` *must* be GPR 0 here.  I think in practice it (nearly?) always is, but it still would be better to at least verify.

elliottt replied:
I'll verify that it's always used with gpr 0, and add a comment explaining why we return it there if that's the case.

uweigand commented on cranelift/codegen/src/isa/s390x/inst.isle:
```diff
@@ -3424,17 +3427,17 @@
+      (mov_preg result))
 (rule (casloop_result (ty_32_or_64 ty) (littleendian) result)
-      (bswap_reg ty result))
+      (bswap_reg ty (mov_preg result)))
```
I'm confused about the exact rules when/which instructions are allowed to refer to non-allocatable fixed registers.  It seems the load before loop (loading into %r0) is OK, but the move is problematic (which I understand is because regalloc handles moves specially).

But is it OK to have a fixed register as source of the bswap instruction?  If so, we wouldn't have to incur this performance regression.  In fact, if *only* moves are the problem, maybe the fix could be as simple as having a special-cased rule for `copy_reg` that detects that the source is a fixed register and uses mov_preg instead?

elliottt replied:
That's a good point: moves are indeed the case that we need to avoid, and it should be fine for the bswap instruction to take a fixed non-allocatable input. I'll fix this up, thanks!

uweigand reviewed (commented):
Thanks for tracking this down!   A couple comments on the PR inline.

elliottt commented:
@cfallin the assertions in `collect_operands` are failing, so I'll open a separate PR with that change and track them down there.

bjorn3 commented:
Will this be backported to the wasmtime-4.0.0 branch?

elliottt commented:
> Will this be backported to the wasmtime-4.0.0 branch?

Just made the PR: #5442
